File: detector.py
import argparse
import cv2
config = None

def config(argv):
    parser = argparse.ArgumentParser()
    subparsers = parser.add_subparsers(
        dest='subparser', required=True
    )
    config_file = subparsers.add_parser('file')
    config_file.add_argument(
        '-path', type=str, action='store',
        help='config file', default='./cascade_config.txt'
    )
    config_args = subparsers.add_parser('args')
    config_args.add_argument(
        '-path', type=str, action='store',
        help='cascade file', required=True
    )
    config_args.add_argument(
        '-min', type=int, action='store',
        help='min window', required=True
    )
    config_args.add_argument(
        '-max', type=int, action='store',
        help='max window', required=True
    )
    config_args.add_argument(
        '-step', type=int, action='store',
        help='window range step', default=30
    )
    config_args.add_argument(
        '-factor', type=float, action='store',
        help='scaleFactor', default=1.5
    )
    ex_gr = config_args.add_mutually_exclusive_group()
    ex_gr.add_argument(
        '-range', action='store_true',
        help='window range step'
    )
    ex_gr.add_argument(
        '-scale', action='store_true',
        help='window range step'
    )
    config_args.add_argument(
        '-base', type=int, action='store',
        help='cascade base window size', default=24
    )
    config_args.add_argument(
        '-neighbors', type=int, action='store',
        help='min num of neighbors', required=True
    )
    stream = {
        'type':str, 'action':'store',
        'help':'url of a video stream of nothing for webcam'
    }
    config_args.add_argument('-stream', **stream)
    config_file.add_argument('-stream', **stream)

    res = parser.parse_args(argv)
    if res.subparser == 'file':
        config = read_file(res.path)
        config['stream'] = res.stream
        return config
    else:
        config = {}
        config['stream'] = res.stream
        config['path'] = res.path
        if res.range:
            config['type'] = 'range'
            config['step'] = res.step
            config['base'] = res.base
        elif res.scale:
            config['type'] = 'scale'
            config['scalefactor'] = res.factor
        config['neighbors'] = res.neighbors
        config['min'] = res.min
        config['max'] = res.max

        return config

def read_file(path):
    config = {}
    with open(path, 'rt') as f:
        i = 0
        for line in f.readlines():
            line = line.strip()
            if len(line) == 0 or line[0] == '#':
                continue
            if i == 0:
                config['path'] = line.strip()
            elif i == 1:
                if line not in ['range', 'scale']:
                    print('wrong config type')
                    import sys
                    sys.exit()
                config['type'] = line
            elif i == 2 or i == 3:
                s = int(line)
                config['min' if i == 2 else 'max'] = s
            elif i == 4:
                if config['type'] == 'range':
                    s = int(line)
                    config['step'] = s
                else:
                    s = float(line)
                    config['scalefactor'] = s
            elif i == 5:
                b = int(line)
                config['base'] = b
            elif i == 6:
                n = int(line)
                config['neighbors'] = n
            i += 1
    return config

def make_detector(config):
    path = config['path']
    cascade = cv2.CascadeClassifier()
    if not cascade.load(path):
        raise Exception("no cascade file")
    return cascade

def detect(image, detector, config):
    gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    rects = []
    if config['type'] == 'range':
        for size in range(config['min'], config['max'], config['step']):
            factor = (size / config['base'])
            trects = detector.detectMultiScale(
                gray, scaleFactor=factor,
                minSize=(size, size), maxSize=(size, size),
                minNeighbors=config['neighbors']
            )
            rects += list(trects)
            del trects
    else:
        min_size = (config['min'],)*2
        max_size = (config['max'],)*2
        rects = detector.detectMultiScale(
            gray, scaleFactor=config['scalefactor'],
            minSize=min_size, maxSize=max_size,
            minNeighbors=config['neighbors']
        )
    return rects

if __name__ == '__main__':
    import sys
    print(config(sys.argv[1:]))

File: detector.cpp
#include "detector.h"

std::ostream& operator<< (std::ostream& os, CascadeConfig const& cc){
    os << "path: " << cc.path << '\n';
    os << "min: " << cc.min << '\n';
    os << "max: " << cc.max << '\n';
    os << "neighbors: " << cc.neighbors << '\n';
    os << "stream: " << cc.stream << '\n';
    os << "step: " << cc.step << '\n';
    os << "base: " << cc.base << '\n';
    os << "scalefactor: " << cc.scalefactor << '\n';
    os << "is_range_type: " << cc.is_range_type << '\n';
    return os;
}


CascadeConfig read_config(std::string path){
    CascadeConfig config;
    std::ifstream ifs(path);
    if(!ifs.is_open()){
        throw std::runtime_error("can not open a cascade config file");
    }
    std::string line;
    for(int i = 0; i < 7; ++i){
        std::getline(ifs, line);
        while(line.size() > 0 && line[0] == '#') std::getline(ifs, line);
        if(i == 0){
            config.path = line;
        }else if(i == 1){
            if (line == std::string("range")){
                config.is_range_type = true;
            }else if(line == std::string("scale")){
                config.is_range_type = false;
            }else{
                std::cout << "wrong config type\n";
                exit(0);
            }
        }else{
            std::istringstream is(line);
            if(i == 2 || i == 3){
                int s;
                is >> s;
                if(i == 2){
                    config.min = s;
                }else{
                    config.max = s;
                }
            }else if(i == 4){
                if(config.is_range_type){
                    int step;
                    is >> step;
                    config.step = step;
                }else{
                    double factor;
                    is >> factor;
                    config.scalefactor = factor;
                }
            }else if(i == 5){
                int base;
                is >> base;
                config.base = base;
            }else{
                int n;
                is >> n;
                config.neighbors = n;
            }
        }
    }

    return config;
}

CascadeConfig config(int argc, char** argv){
    CascadeConfig res;
    cv::CommandLineParser parser(argc, argv,
        "{h help      |                      |}"
        "{file        |                      |}"
        "{args        |                      |}"
        "{stream      |                      |}"
        "{min         | 100                  |}"
        "{max         | 160                  |}"
        "{neighbors   | 3                    |}"
        "{step        | 30                   |}"
        "{base        | 24                   |}"
        "{path        |                      |}"
        "{type        | scale                |}"
        "{factor      | 1.5                  |}"
    );
    if (parser.has("h")){
        parser.printMessage();
        exit(0);
    }
    std::string stream = parser.get<std::string>("stream");
    if(parser.has("file")){
        std::string config_path = parser.get<std::string>("file");
        if (config_path == "true"){
            config_path = "./cascade_config.txt";
        }
        res = read_config(config_path);
        res.stream = stream;
    }else if(parser.has("args")){
        res.stream = stream;
        res.neighbors = parser.get<int>("neighbors");
        if(!parser.has("path")){
            std::cout << "path is needed\n";
            exit(0);
        }
        res.path = parser.get<std::string>("path");
        res.min = parser.get<int>("min");
        res.max = parser.get<int>("max");
        res.step = parser.get<int>("step");
        res.base = parser.get<int>("base");
        res.scalefactor = parser.get<double>("factor");
        if(parser.get<std::string>("type") == "range"){
            res.is_range_type = true;
        }else if(parser.get<std::string>("type") == "scale"){
            res.is_range_type = false;
        }else{
            std::cout << "wrong config type\n";
            exit(0);
        }

        // int w, h;
        // std::istringstream miniss(parser.get<std::string>("min"));
        // miniss >> w;
        // miniss.ignore(1);
        // miniss >> h;
        // res.min = cv::Size(w, h);
        // std::istringstream maxiss(parser.get<std::string>("max"));
        // maxiss >> w;
        // maxiss.ignore(1);
        // maxiss >> h;
        // res.max = cv::Size(w, h);
    }else{
        std::cout << "either args or file is needed\n";
        exit(0);
    }
    std::cout << res;
    return res;
}

cv::CascadeClassifier make_detector(CascadeConfig& cascade_config){
    cv::CascadeClassifier cascade;
    if(!cascade.load(cascade_config.path)){
        throw std::runtime_error("no cascade file");
    }
    return cascade;
}

std::vector<cv::Rect> detect_objects(
    cv::Mat& image, cv::CascadeClassifier& detector, CascadeConfig& cascade_config
){
    cv::Mat gray;
    cvtColor(image, gray, cv::COLOR_BGR2GRAY);
    if (cascade_config.is_range_type){
        std::vector<cv::Rect> objects;
        for(
            int size = cascade_config.min;
            size <= cascade_config.max;
            size += cascade_config.step
        ){
            double factor = size / cascade_config.base;
            cv::Size min(size, size);
            cv::Size max(size, size);
            std::vector<cv::Rect> rects;
            detector.detectMultiScale(
                gray, rects, factor,
                cascade_config.neighbors, 0,
                min, max
            );
            objects.insert(end(objects), begin(rects), end(rects));
        }
        return objects;
    }else{
        std::vector<cv::Rect> objects;
        cv::Size min(cascade_config.min, cascade_config.min);
        cv::Size max(cascade_config.max, cascade_config.max);
        detector.detectMultiScale(
            gray, objects, cascade_config.scalefactor,
            cascade_config.neighbors, 0,
            min, max
        );
        return objects;
    }
}
